File: frontend/src/api/knowledge.ts
/** 知识库模块接口（B2：KB / 文档 / Chunk）。 */

import { client, request } from './client';

export interface KnowledgeBase {
  id: string;
  name: string;
  description: string;
  doc_count: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeBaseDetail {
  id: string;
  name: string;
  description: string;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentItem {
  id: string;
  kb_id: string;
  file_name: string;
  file_type: string;
  file_size: number;
  status: 'uploading' | 'parsing' | 'embedding' | 'completed' | 'failed';
  error_message: string | null;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface ChunkItem {
  id: string;
  doc_id: string;
  kb_id: string;
  chunk_index: number;
  question: string;
  answer: string;
  category: string | null;
  page: string | null;
  row: string | null;
  word_count: number;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface PageResult<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export interface ChunkPayload {
  doc_id?: string;
  question: string;
  answer: string;
  category?: string | null;
  tags?: string[];
  page?: string | null;
  row?: string | null;
}

// ---------- 知识库 ----------

export function listKnowledgeBases(): Promise<KnowledgeBase[]> {
  return request<KnowledgeBase[]>({ url: '/knowledge-bases', method: 'GET' });
}

export function getKnowledgeBase(id: string): Promise<KnowledgeBaseDetail> {
  return request<KnowledgeBaseDetail>({ url: `/knowledge-bases/${id}`, method: 'GET' });
}

export function createKnowledgeBase(data: {
  name: string;
  description: string;
}): Promise<KnowledgeBase> {
  return request<KnowledgeBase>({
    url: '/knowledge-bases',
    method: 'POST',
    data,
  });
}

export function updateKnowledgeBase(
  id: string,
  data: { name: string; description: string },
): Promise<KnowledgeBase> {
  return request<KnowledgeBase>({
    url: `/knowledge-bases/${id}`,
    method: 'PUT',
    data,
  });
}

export function deleteKnowledgeBase(id: string): Promise<null> {
  return request<null>({ url: `/knowledge-bases/${id}`, method: 'DELETE' });
}

// ---------- 文档 ----------

export function listDocuments(
  kbId: string,
  params: { keyword?: string; page: number; page_size: number },
): Promise<PageResult<DocumentItem>> {
  return request<PageResult<DocumentItem>>({
    url: `/knowledge-bases/${kbId}/documents`,
    method: 'GET',
    params,
  });
}

export function getDocument(docId: string): Promise<DocumentItem> {
  return request<DocumentItem>({ url: `/documents/${docId}`, method: 'GET' });
}

export function deleteDocument(docId: string): Promise<null> {
  return request<null>({ url: `/documents/${docId}`, method: 'DELETE' });
}

export function reparseDocument(docId: string): Promise<DocumentItem> {
  return request<DocumentItem>({ url: `/documents/${docId}/reparse`, method: 'POST' });
}

export async function uploadDocument(
  kbId: string,
  file: File,
  onProgress?: (percent: number) => void,
): Promise<{ document_id: string; file_name: string; status: string }> {
  const form = new FormData();
  form.append('file', file);
  const resp = await client.post(`/knowledge-bases/${kbId}/documents`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
    onUploadProgress: (e) => {
      if (e.total) {
        onProgress?.(Math.round((e.loaded / e.total) * 100));
      }
    },
  });
  return resp.data.data;
}

// ---------- Chunk ----------

export function listChunks(
  docId: string,
  params: { page: number; page_size: number },
): Promise<PageResult<ChunkItem>> {
  return request<PageResult<ChunkItem>>({
    url: `/documents/${docId}/chunks`,
    method: 'GET',
    params,
  });
}

export function createChunk(payload: ChunkPayload): Promise<ChunkItem> {
  return request<ChunkItem>({ url: '/chunks', method: 'POST', data: payload });
}

export function updateChunk(id: string, payload: Partial<ChunkPayload>): Promise<ChunkItem> {
  return request<ChunkItem>({ url: `/chunks/${id}`, method: 'PUT', data: payload });
}

export function deleteChunk(id: string): Promise<null> {
  return request<null>({ url: `/chunks/${id}`, method: 'DELETE' });
}

File: backend/app/schemas/knowledge.py
"""知识库 / 文档 / Chunk 请求与响应模型（08 §5.2 / §6.2）。"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


def _clean_tags(tags: list[str] | None) -> list[str]:
    """标签清洗：去空白、去重、单标签 ≤20 字、最多 10 个（04 §4.5 新增规则）。"""
    if not tags:
        return []
    cleaned: list[str] = []
    for tag in tags:
        tag = (tag or "").strip()
        if not tag:
            continue
        tag = tag[:20]
        if tag not in cleaned:
            cleaned.append(tag)
    return cleaned[:10]


class KnowledgeBaseCreate(BaseModel):
    name: str = Field(min_length=1, max_length=50, description="名称（唯一，≤50 字）")
    description: str = Field(default="", max_length=200, description="描述（≤200 字）")


class KnowledgeBaseUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=50)
    description: str = Field(default="", max_length=200)


class KnowledgeBaseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str
    doc_count: int = 0
    created_at: datetime
    updated_at: datetime


class KnowledgeBaseDetailOut(BaseModel):
    """详情在列表字段基础上附加创建人。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str
    created_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kb_id: uuid.UUID
    file_name: str
    file_type: str
    file_size: int
    status: str
    error_message: str | None = None
    chunk_count: int
    created_at: datetime
    updated_at: datetime


class UploadDocumentOut(BaseModel):
    document_id: uuid.UUID
    file_name: str
    status: str


class ChunkCreate(BaseModel):
    doc_id: uuid.UUID
    question: str = Field(min_length=1, max_length=200, description="问题（必填，≤200 字）")
    answer: str = Field(min_length=1, max_length=2000, description="答案（必填，≤2000 字）")
    category: str | None = Field(default=None, max_length=50, description="分类")
    tags: list[str] = Field(default_factory=list, description="标签（最多 10 个，单个 ≤20 字）")
    page: str | None = Field(default=None, max_length=50)
    row: str | None = Field(default=None, max_length=50)

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, v: list[str]) -> list[str]:
        return _clean_tags(v)

    @field_validator("category")
    @classmethod
    def strip_category(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        return v[:50] or None


class ChunkUpdate(BaseModel):
    question: str | None = Field(default=None, min_length=1, max_length=200)
    answer: str | None = Field(default=None, min_length=1, max_length=2000)
    category: str | None = Field(default=None, max_length=50)
    tags: list[str] | None = Field(default=None)
    page: str | None = Field(default=None, max_length=50)
    row: str | None = Field(default=None, max_length=50)

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, v: list[str] | None) -> list[str] | None:
        return _clean_tags(v) if v is not None else None


class ChunkOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    doc_id: uuid.UUID
    kb_id: uuid.UUID
    chunk_index: int
    question: str
    answer: str
    category: str | None = None
    page: str | None = None
    row: str | None = None
    word_count: int
    tags: list[str]
    created_at: datetime
    updated_at: datetime

